File: commons/src/main/java/org/dhis2/commons/data/tuples/Quintet.java
package org.dhis2.commons.data.tuples;

import androidx.annotation.NonNull;

import com.google.auto.value.AutoValue;

@AutoValue
public abstract class Quintet<A, B, C, D, E> {

    @NonNull
    public abstract A val0();

    @NonNull
    public abstract B val1();

    @NonNull
    public abstract C val2();

    @NonNull
    public abstract D val3();

    @NonNull
    public abstract E val4();

    @NonNull
    public static <A, B, C, D, E> Quintet<A, B, C, D, E> create(@NonNull A val0,
                                                                @NonNull B val1, @NonNull C val2,
                                                                @NonNull D val3, @NonNull E val4) {
        return new AutoValue_Quintet<>(val0, val1, val2, val3, val4);
    }
}

File: app/src/main/java/org/dhis2/data/tuples/Quintet.kt
package org.dhis2.data.tuples

import com.google.auto.value.AutoValue

@AutoValue
abstract class Quintet<A, B, C, D, E> {

    abstract fun val0(): A

    abstract fun val1(): B

    abstract fun val2(): C

    abstract fun val3(): D

    abstract fun val4(): E

    companion object {

        fun <A, B, C, D, E> create(val0: A,
                                   val1: B, val2: C,
                                   val3: D, val4: E): Quintet<A, B, C, D, E> {
            return AutoValue_Quintet(val0, val1, val2, val3, val4)
        }
    }
}
